### ISSUE TITLE
Support overriding the implementation of a federated plugin

### ISSUE THREAD

**Issue by IchordeDionysos** [flutter/flutter#59657]
## Use case

For devices, without GMS we want to override the default implementation of packages like `firebase_auth`.
To achieve this we want to implement the `firebase_auth_platform_interface` and use our custom class.

## Proposal

In the federated plugin proposal by @amirh, there is a section about overriding the default implementation with a custom package:
https://docs.google.com/document/d/1LD7QjmzJZLCopUrFAAE98wOUQpjmguyGTN2wd_89Srs/edit#heading=h.zc9ie62xv82k

We thought something like this should be possible:

pubspec.yaml
```
dependencies: 
  firebase_auth: any
  firebase_auth_android_rest: any
```

And because the Flutter tool knows that `firebase_auth_android_rest` provides a custom implementation of the `firebase_auth_platform_interface`/`firebase_auth` package, it uses `firebase_auth_android_rest` and not `firebase_auth` for Android.

We don't know if this is possible yet.

**Comment by jonahwilliams** [comment]
This can already be done via pub `dependency_overrides`:  https://dart.dev/tools/pub/dependencies#dependency-overrides

**Comment by creativecreatorormaybenot** [comment]
@jonahwilliams No, it cannot - unfortunately.

What this issue is referring to is overriding implementations of federated plugins. The main `pubspec.yaml` of a federated package specifies the packages that create the platform implementations.

This means that if you want to only override say the Android implementation of a federated plugin, you **need to** override the main package as well.

**Comment by jonahwilliams** [comment]
> What this issue is referring to is overriding implementations of federated plugins. The main pubspec.yaml of a federated plugin specifies the packages that create the platform implementations.

Yes, you can override dependencies of dependencies

**Comment by creativecreatorormaybenot** [comment]
@jonahwilliams Of course, but you cannot supply a *different* package for federated plugins (which is what this issue attempts to describe).

Our use case is the following: we created the [`firebase_rest_auth`](https://pub.dev/packages/firebase_rest_auth) plugin and this is supposed to be an **Android replacement** for the `firebase_auth` plugin.

If you take a look at the [installing guide](https://github.com/simpleclub/firebase_rest_auth#installation), you will hopefully see what I mean 🙂  

We want to offer a Pub package that only replaces a platform implementation of a ***federated plugin***, however, that is currently not possible without overriding the actual plugin package.  
Even if we wanted to override the web implementation, which is in a separate package by default, we cannot do that with a *different* package.

The only possible way would be a dependency override of `firebase_auth_web`, however, we **cannot** publish that on Pub because that obviously already exists.

The section in the document @IchordeDionysos linked seems to indicate that replacing platform implementations should be supported - that is what this issue is about.

---

I hope I was able to explain the issue better now 😄
